lib/collection/src/collection/resharding.rs

```diff
@@ -162,30 +181,31 @@ impl Collection {
         let _ = self.stop_resharding_task(&resharding_key).await;
         shard_holder.finish_resharding_unchecked(&resharding_key)?;
 
-        // Update the stored shard count, ensures we load the new shard on restart
+        Ok(())
+    }
+
+    pub async fn abort_resharding(&self, resharding_key: ReshardKey) -> CollectionResult<()> {
+        let _ = self.stop_resharding_task(&resharding_key).await;
+
+        // Decrease the persisted shard count, ensures we don't load dropped shard on restart
         {
             let mut config = self.collection_config.write().await;
-            config.params.shard_number = config.params.shard_number.saturating_add(1);
 
-            // If using automatic sharding, the reshard key shard ID must correspond with number
+            #[cfg(debug_assertions)]
             if config.params.sharding_method.unwrap_or_default() == ShardingMethod::Auto {
-                debug_assert_eq!(
+                assert_eq!(
                     config.params.shard_number.get() - 1,
-                    resharding_key.shard_id,
+                    resharding_key.shard_id
                 );
             }
 
+            config.params.shard_number = NonZeroU32::new(config.params.shard_number.get() - 1)
+                .expect("cannot have zero shards after aborting resharding");
```

Comment:
That does not work.

It is a `NonZeroU32` which does not have [`_sub`](https://doc.rust-lang.org/std/num/type.NonZeroU32.html) methods. Calling `checked_sub(1)` on the raw `u32` does not make sense, because it will always be 1 or more.